crates/re_viewport/src/context_menu/mod.rs

```diff
@@ -69,232 +77,219 @@ pub fn context_menu_ui_for_item(
+fn show_context_menu_for_selection(ctx: &ContextMenuContext<'_>, ui: &mut egui::Ui) {
+    let mut should_display_separator = false;
+    for action_section in action_list(ctx.viewer_context) {
+        let mut any_action_displayed = false;
+
+        for action in action_section {
+            if !action.supports_selection(ctx) {
+                continue;
+            }
+
+            any_action_displayed = true;
+
+            if should_display_separator {
+                ui.separator();
+                should_display_separator = false;
+            }
 
-        for action in actions {
-            let response = action.ui(ctx, viewport_blueprint, ui);
+            let response = action.ui(ctx, ui);
             if response.clicked() {
                 ui.close_menu();
             }
         }
-    });
-}
-
-// ---
 
-/// Trait for things that can populate a context menu
-trait ContextMenuItem {
-    // TODO(ab): return a `ListItem` to make those context menu nice to look at. This requires
-    // changes to the context menu UI code to support full-span highlighting.
-    fn label(&self, _ctx: &ViewerContext<'_>, _viewport_blueprint: &ViewportBlueprint) -> String {
-        String::new()
+        should_display_separator |= any_action_displayed;
     }
+}
 
-    fn run(&self, _ctx: &ViewerContext<'_>, _viewport_blueprint: &ViewportBlueprint) {}
+/// Context information provided to context menu actions
+struct ContextMenuContext<'a> {
+    viewer_context: &'a ViewerContext<'a>,
+    viewport_blueprint: &'a ViewportBlueprint,
+    selection: &'a Selection,
+    clicked_item: &'a Item,
+}
 
-    /// run from inside of [`egui::Response.context_menu()`]
-    fn ui(
-        &self,
-        ctx: &ViewerContext<'_>,
-        viewport_blueprint: &ViewportBlueprint,
-        ui: &mut egui::Ui,
-    ) -> egui::Response {
-        let label = self.label(ctx, viewport_blueprint);
-        let response = ui.button(label);
-        if response.clicked() {
-            self.run(ctx, viewport_blueprint);
+/// Context menu actions must implement this trait.
+///
+/// Actions must do three things, corresponding to three core methods:
+/// 1. Decide if it can operate a given [`Selection`] ([`Self::supports_selection`]).
+/// 2. If so, draw some UI in the context menu ([`Self::ui`]).
+/// 3. If clicked, actually process the [`Selection`] ([`Self::process_selection`]).
+///
+/// For convenience, these core methods have default implementations which delegates to simpler
+/// methods (see their respective docstrings). Implementor may either implement the core method for
+/// complex cases, or one or more of the helper methods.
+trait ContextMenuAction {
+    /// Check if the action is able to operate on the provided selection.
+    ///
+    /// The default implementation delegates to [`Self::supports_multi_selection`] and
+    /// [`Self::supports_item`].
+    fn supports_selection(&self, ctx: &ContextMenuContext<'_>) -> bool {
```

Comment:
kinda surprised there's only support and not a disable/enabled option. I suppose this could be later extended to return an enum that distinguises this

Reply:
I'm not sure what you refer to by disabled/enabled? Bi-stable actions (like "Hide/Show"—which is now split into two separate actions btw)? Or actually displaying the action as disabled in the context menu (as opposed to no showing it at all)?

Reply:
it looks like this hasn't come up yet, but I'd expect us to very soon have items that are displayed but disabled/greyed out because of some reason

Reply:
I guess. Well, it'll be easy to switch to an enum then :)